data/TAM/sai_tam_port.proto

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2017 Broadcom. The term "Broadcom" refers
+ * to Broadcom Limited and/or its subsidiaries.
+ 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ 
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+syntax = "proto2";
+import "sai_tam_main.proto";
+
+extend SAISwitchSensors {
+    optional GPort sai_interface_ext = 3;
+}
+
+message GPort {
+    repeated InterfaceInfos interface_stats = 1;
+}
+            
+message InterfaceInfos {
+    required string if_name                = 1 [(telemetry_options).is_key = true];
+    optional uint64 init_time              = 2;     // time when if/stats last reset
+
+    // Internal Port ID
+    optional uint32 if_pid                 = 3;
+
+    // SNMP Index
+    optional uint32 snmp_if_index          = 4;
+
+    // name of parent for lag interface, if applicable
+    optional string parent_lag_name        = 5;
+
+    // ingress traffic stats
+    optional InterfaceStats ingress_stats  = 6;
+
+    // egress traffic stats
+    optional InterfaceStats egress_stats   = 7;
+
+    // ingress traffic errors
+    optional InterfaceErrors ingress_errors = 8;
+
+    // egress traffic errors
+    optional InterfaceErrors egress_errors = 9;
+
+    // Interface administration status
+    optional string if_admin_status        = 10;
+
+    // Interface operational status
+    optional string if_op_status           = 11;
+
+    // number of carrier transitions on this interface
+    optional uint64 if_transitions         = 12 [(telemetry_options).is_counter = true]; 
+
+    // This corresponds to the ifLastChange object in the standard interface MIB
+    optional uint32 ifLastChange           = 13;
+
+    // This corresponds to the ifHighSpeed object in the standard interface MIB
+    optional uint32 ifHighSpeed            = 14;
+
+    repeated QueueStats egress_queue_info  = 15;    // egress queue information
+    repeated QueueStats ingress_queue_info = 16;    // ingress queue information
```

Comment:
Quoting from Jai's reply:

> [Jai] message is extensible. More types can be added as optional. This is same in proto3 as well. Think it as union in C (without the datatype width)

Reply:
In most if not all of the enums specifying attributes in the SAI .h files, including all in saitam.h, a range of "custom" attributes are defined at the end. This allows for custom or experimental functionality to be developed, without churning the base .h files and without requiring explicit approval from the OCP SAI subproject.

proto2 extensions have the same purpose, provide all the same benefits, and work in exactly the same way. A developer of custom or experimental functionality should be able to extend the Protobuf message without modifying the mainstream .proto file. proto2 extensions (or proto3 Any) enable this. With the use of extensions, custom or experimental .proto files can be defined that extend this message, without churning the base .proto files.

All that has to be done is adding at the end of each message:
"    extensions 100 to 1000;"
See https://developers.google.com/protocol-buffers/docs/proto#extensions for details.

Why wouldn't we provide the same extensibility in the Protobuf messages that we define for reporting telemetry, as we do in the SAI objects and attributes defined for configuring telemetry?

Consider the following scenario:
// This is the current proposal
message InterfaceInfos {
  // ...
  optional QueueStats ingress_queue_info = 17;
}
// Experimental feature X adds a field, making its own copy of the .proto file,
// which is bad - it should be defined in a separate .proto file.
message InterfaceInfos {
  // ...
  optional QueueStats ingress_queue_info = 17;
  optional VendorXType vendorX_field = 18;
}
// Now a new field is added to the official OCP TAM proto
message InterfaceInfos {
  // ...
  optional QueueStats ingress_queue_info = 17;
  optional SomeType official_field = 18;
}
Backward compatibility is broken and all hell breaks loose.
But all you needed to do was include this line in your original message:
message InterfaceInfos {
  // ...
  optional QueueStats ingress_queue_info = 17;
  extensions 100 to 1000;
}

Reply:
Quoting from Jai's reply:

> If you look into sai_tam_main.proto:: message(SAISwitchSensors)

> It has the message defined as an extensions.


> message SAISwitchSensors {

>     extensions 1 to max;

> }

> What this means is that if a vendor wants to define a custom message, it can root it at SAISwitchSensors level and create a custom hierarchy. This is allowed.

> Suggestion you are making is that what if there is a only a new field which need to be added by specific vendor.

> Adding a new field is also allowed but proposal is that vendor MUST publish the data field as an optional field in the existing message.

> Problem with your suggestion is following

> Multiple vendors may define same extension value for a given data type. This will create a collision and leave the proto file unusable, unless…

> Unless vendor publishes that extension in the SAI proto file and make it visible to all.

> If it MUST publish the data field then making it as part of the message and keeping it optional is a better idea instead of forcing a range in the extension thereby limiting the extensions to a range.


> Hope it explains it.

Reply:
The current definition allows custom or experimental messages to be added as extensions. However, it does not allow the addition of custom or experimental fields in messages such as QueueStats, instead it requires new fields to be defined as full optional fields. While it is true that proto2 extensions can have collisions between different extensions, at least they prevent collisions between extensions and future additions to the base message. Disallowing extensions due to collisions between extensions is throwing out the baby with the bathwater. If you really want to avoid collisions between extensions, move to proto3 and use Any, which solves exactly this issue.

Note also that everything you are arguing is 100% applicable in exactly the same way to the SAI .h files. Why do we have custom ranges in enums defining attributes? Those custom ranges in the enums defining attributes are just as susceptible to collisions between different custom or experimental definitions as proto2 extensions, yet the community still felt that there is sufficient value and defined the custom ranges all over the place.